incorporate findsat_mrt into acstools

dvstark opened the pull request:
<!-- This comments are hidden when you submit the pull request,
so you do not need to remove them! -->

<!-- Please be sure to check out our code of conduct,
https://github.com/spacetelescope/acstools/blob/master/CODE_OF_CONDUCT.md . -->

<!-- Please just have a quick search on GitHub to see if a similar
pull request has already been posted.
We have old closed pull requests that might provide useful code or ideas
that directly tie in with your pull request. -->

<!-- We have several automatic features that run when a pull request is open.
They can appear daunting but do not worry because maintainers will help
you navigate them, if necessary. -->

### Description
<!-- Provide a general description of what your pull request does.
Complete the following sentence and add relevant details as you see fit. -->

<!-- In addition please ensure that the pull request title is descriptive. -->

This pull request is incorporate newcode (findsat_mrt) into acstools. This code, described in ISR 2022-08, provides a new framework for identifying and masking satellite trails in ACS/WFC data. It should not impact any existing acstools programs.

<!-- If the pull request closes any open issues you can add this.
If you replace <Issue Number> with a number, GitHub will automatically link it.
If this pull request is unrelated to any issues, please remove
the following line. -->

dvstark commented:
I'm aware of the odd line breaks with ^M at the end, and am trying to fix it

dvstark commented:
end line character issue fixed. Code should now display normally

jryon reviewed (changes requested):
Looking pretty good to me, but some changes are needed. I've left inline comments on specific items, but it's very likely I haven't covered everything.

One general concern I have is that it looks like you are making a lot of diagnostic plots, which I think is great, but the way it is written assumes users have interactive plotting enabled. Not sure we can/should assume that, so I'd suggest instead to save to disk every time a plotting function is called.

Looks like a typo in one of the doc filenames: `doc/source/utils_findsat_mrt.py` should be `doc/source/utils_findsat_mrt.rst`

Thanks for your hard work to integrate this into acstools!

dvstark commented on acstools/findsat_mrt.py:
```diff
@@ -0,0 +1,1129 @@
+Point sources are then extracted from the MRT and filtered to yield a final 
+catalog of trails. These trails can then be used to create a mask.
+
+Example 1: Identificaation of trails in an ACS/WFC image, j97006j5q_flc.fits (4th extension)
```
Regarding the interactive plotting, would it be ok to have python check if interactive mode is running, e.g.:
`import matplotlib as mpl
mpl.is_interactive()`
If that returns false, the plots can be written out to file, otherwise, they could appear as popups (I could also add an option to both plot on screen and save to disk)

jryon replied:
This solution sounds good to me.

dvstark commented on acstools/findsat_mrt.py:
```diff
@@ -0,0 +1,1129 @@
+        if (np.any(image) != None) & (self.plot == True):
+            self.plot_image()
+
+    def run_mrt(self,theta=None,threads=None):
```
The idea here was flexibility and it being clear which attributes from trailfinder go into which functions. For instance, without the theta and threads keywords, it may not be obvious that these are things that can be adjusted. Is there a best practice here? If possible, I'd like to keep the keyword options, but if they end up being used, I could then update the relevant attributes of the satfinder class.

jryon replied:
I searched around a bit, but couldn't find a specific best practice for this kind of thing. I'd be fine with keeping the keywords and updating the attributes if they are used.

dvstark commented on acstools/findsat_mrt.py:
```diff
@@ -0,0 +1,1129 @@
+            ax=self.plot_mrt()
+            for s in self.source_list:
+                ax.scatter(s['xcentroid'],s['ycentroid'],edgecolor='red',facecolor='none',s=100,lw=2)
+
```
Added warning if no sources found.

dvstark commented on acstools/findsat_mrt.py:
```diff
@@ -0,0 +1,1129 @@
+            
+            
+    def add_streak(self,endpoints,flux,width=3,psf_sigma=None):
+        '''
```
Right, it is no longer needed so I removed it.

dvstark commented on acstools/findsat_mrt.py:
```diff
@@ -0,0 +1,1129 @@
+        
+
+def create_mrt_line_kernel(width,sigma,outfile=None,shape=(1024,2048),plot=False,theta=np.arange(0,180,0.5),threads=1):
+    '''
```
moved it as suggested

pllim commented on .gitattributes:
```diff
@@ -0,0 +1,14 @@
+*.h text
+
+# Declare files that will always have CRLF line endings on checkout.
+*.sln text eol=crlf
```
Is this Windows specific?

dvstark replied:
Not sure. I was having issues with github not understand line breaks in my files and this was a solution that seemed to work. I'm not on Windows, but Mac. Better way to accomplish this?

pllim replied:
What is `*.sln`?

dvstark replied:
I think I copied/pasted a "solution" to my problem without fully understanding, and my .gitattributes files actually may not be necessary. I removed it and the files seem to not have any linebreak problems. Anymore. I think this is resolved.

pllim reviewed (commented):
tl;dr -- How does this relate to the existing satdet? Should that one be deprecated or something?

New doc pages not linked from top level index page?

There are no tests at all. How do we know code is working correctly and how to prevent future regression?

pllim commented:
Also, CI failures need to be addressed.

dvstark commented on acstools/findsat_mrt.py:
```diff
@@ -0,0 +1,1129 @@
+            if self.source_list is not None:
+                self.source_list.write('{}/{}_catalog.fits'.format(output_dir,root),overwrite=True)
+            else:
+                #create an empty catalog and write that. It helps to have this for future analysis purposes even if it's empty
```
I let an empty catalog be written because it made it clear that the image was inspected but nothing found, as opposed to an image that was never inspected. If I did not write a catalog for images where no trails were found, I could make a note in the log, but I feel like this adds another layer of complexity to analysis downstream when merging all the catalog files. With that explanation, is it ok to leave the code as is? Happy to discuss alternatives!

dvstark commented on acstools/utils_findsat_mrt.py:
```diff
@@ -0,0 +1,1047 @@
+            counter += 1
+        else:
+            #check that added sources are not recreations of already detected ones
+            keep = (np.zeros(len(t))+1).astype(bool)
```
this should be fine. It checks each source and sets the value to keep=false if its too close to an existing source

pllim commented on acstools/tests/test_wfc_findsat_mrt.py:
```diff
@@ -0,0 +1,39 @@
+
+from acstools.findsat_mrt import wfc_wrapper
+from acstools.tests.helpers import BaseACSTOOLS
+import acstools as a
+package_path = a.__path__[0]
```
Not sure what this is doing.
```suggestion
```

pllim commented on acstools/tests/test_wfc_findsat_mrt.py:
```diff
@@ -0,0 +1,39 @@
+
+#update all this with my code
+
+class test_findsat_mrt(BaseACSTOOLS):
```
```suggestion
class TestFindsatMRT(BaseACSTOOLS):
```

pllim commented on acstools/tests/test_wfc_findsat_mrt.py:
```diff
@@ -0,0 +1,39 @@
+
+        wfc_wrapper(inputfile, binsize=4, extension=4,
+                    output_root='jc8m32j5q',
+                    output_dir = package_path + '/tests/data/input/',
```
Use tmp_path if possible: https://docs.pytest.org/en/7.1.x/how-to/tmp_path.html

dvstark replied:
Prefer to leave as is for the following reason: I'm trying to use the "compare_outputs" code which appears to only look in the /data/input and /data/truth/ directories for the files to test. So it is easier if my code writes the output files to the /data/input/ directory.

pllim replied:
Do you really need `compare_outputs` though. I haven't looked at your code in depth but that test class was written for HSTCAL that only can write to "working directory".

Anyways, if you are adamant on this, you have to make sure Git ignore those output files so you don't accidentally check them into version control.

dvstark replied:
I modified the test to not use compare_outputs and use the temporary directory for the output files.

pllim commented on acstools/tests/test_wfc_findsat_mrt.py:
```diff
@@ -0,0 +1,39 @@
+
+class test_findsat_mrt(BaseACSTOOLS):
+    detector = 'wfc'
+    package_directory = os.path.dirname(os.path.abspath(__file__))
```
```suggestion
```

dvstark commented:
At this point I believe I've addressed all comments and it passed all the automated checks. Some final responses to points below:

> tl;dr -- How does this relate to the existing satdet? Should that one be deprecated or something?

It's an alternative method, with some key advantages (e.g., sensitivity), but it also performs poorly under certain situations (very crowded field). It's possible satdet works better in crowded fields (although this has not been thoroughly tested), and so we prefer to leave it in place. It's up to the use which they want to use. 

> New doc pages not linked from top level index page?

Should be fixed now.

> There are no tests at all. How do we know code is working correctly and how to prevent future regression?

I added a test.

pllim reviewed (commented):
Formatted HTML looks a little weird: https://acstools--176.org.readthedocs.build/en/176/findsat_mrt.html

In general, docstring should follow https://numpydoc.readthedocs.io/en/latest/format.html style guide. Section headings should be consistent throughout the package (see other modules for what level should you start at). "Hidden" functions (those with leading underscore) should not be exposed in API doc unless there is a good reason.

Code snippet continuation should use "..." and not ">>>".

pllim commented on doc/source/utils_findsat_mrt.py:
```diff
@@ -0,0 +1,9 @@
+
+**************************
+MRT-based Satellite Trails Detection Utilities
+**************************
```
Section underline should match the number of characters.

dvstark replied:
This file was actually out of date and should have been removed from the repository. Now done.

pllim commented on setup.cfg:
```diff
@@ -79,6 +83,8 @@ filterwarnings =
   # This is from skimage import
   ignore:.*np\.bool8.*is a deprecated alias:DeprecationWarning
+  #often appears during MRT calculation but is inconsequential
+  ignore:All-NaN slice encountered:RuntimeWarning
```
I feel like if you don't care about such warning, it should be handled properly during the calculation in the code, not as blanket ignore for test suite that also affect other modules. Please have a look to see if `np.seterr` is what you really want and use it as context manager to the affected calculation only.

dvstark replied:
I now use warnings.filterwarnings for that specific warning, only inside the trailfinder class. It's very common due to how missing data are presented in the module, but it is inconsequential.

pllim commented on acstools/tests/test_wfc_findsat_mrt.py:
```diff
@@ -0,0 +1,54 @@
+
+    detector = 'wfc'
+
+    def test_wfc_wrapper(self, tmpdir):
```
If you want to future-proof, `tmpdir` is now considered "legacy" and `tmp_path` should be used instead. See https://docs.pytest.org/en/stable/how-to/tmp_path.html#tmpdir-and-tmpdir-factory

dvstark replied:
Response: I have replaced tmpdir with tmp_path

pllim commented on acstools/utils_findsat_mrt.py:
```diff
@@ -0,0 +1,1253 @@
+__author__ = "David V. Stark"
+__version__ = "1.0"
+__vdate__ = "06-Dec-2022"
+__all__ = ['_round_up_to_odd', 'merge_tables', 'good_indices',
```
Why include the "hidden" functions?

dvstark replied:
Didn't know the standard here. They have been removed, although I have generally modified which are "hidden".

pllim commented on setup.cfg:
```diff
@@ -35,6 +35,8 @@ install_requires =
     astropy>=3.1
     requests
     PyYAML
+    photutils
```
If you put these in `install_requires`, they are not needed in `extras_require`.

You should think about whether these are really required at install time (`install_requires`) or they are merely optional dependencies that are only needed if one needs to use one small part of the package (`extras_require`).

pllim replied:
Optional dependencies should only be imported locally when needed and nothing should crash horribly outside of the things that actually use them when they are not present.

dvstark replied:
Given that these are used at only specific times in the pipeline, I moved them to extras_require.

pllim replied:
I still see them listed here. Did you forget to remove?

dvstark replied:
Yes, I just forgot to commit. Fixed

pllim commented on doc/source/index.rst:
```diff
@@ -40,6 +40,8 @@ To install the development version of this package::
    acszpt
    acsphotcte
    satdet
+   findsat_mrt
+   utils_findsat_mrt
```
Why can't MRT only have one page?

dvstark replied:
It can. I modified the rst files. utils_findsat_mrt.rst is no longer needed and has been deleted

pllim commented on acstools/utils_findsat_mrt.py:
```diff
@@ -0,0 +1,1253 @@
+    image = add_streak(image, width, 1, rho=0, theta=90, psf_sigma=sigma)
+
+    # plot the image
+    if (plot is True) & (plt is not None):
```
Why are two conditions even needed for this check? 
```suggestion
    if plt is not None:
```

dvstark replied:
Response: "plt" says whether matplotlib exists, while plot says whether plots should be generated. I don't think those should necessarily be  merged into a single flag.

pllim commented on acstools/utils_findsat_mrt.py:
```diff
@@ -0,0 +1,1253 @@
+    # the max.
+    rt_rho = np.nansum(rt, axis=1)
+    rt_theta = np.nansum(rt, axis=0)
+    fig, [ax1, ax2] = plt.subplots(1, 2)
```
Won't the rest of this function crash horribly from here onwards if you don't have matplotlib?

dvstark replied:
Yes, thanks for catching that. Those commands should be inside an if statement

dvstark commented:
I have examined the docstring formatting link, as well as the STScI style guides, made the following general changes:
(1) Ensured all routines have single line description at the top. Any additional documentation comes after separated by an additional blank line
(2) replace >>> with ... for line continuations
(3) Adjusted headings around "Examples" to match other acstools routines (although existing docs don't always seem consistent)
(4) Took greater care when defining "hidden" functions (with _ before them). I may have not been doing this properly, so there are only very few of these now.
(5) More in-line commenting throughout
(6) Updated some class and variable names to be consistent with STScI style guide
(7) Not really a style issue, but I also added a routine to update the DQ array in an input flc/flt file, identically to how satdet does it.

I have tried to make the code match best standards. I am getting a new failure in CI / tests and CI / dev_deps_tests, which may be related to the issue discussed here: https://github.com/Lightning-AI/lightning/issues/16756. I'd appreciate someone more familiar with github workflows to take a look.

pllim commented:
Re: failing CI -- I ignored the warning in `master` branch, so if you rebase, the CI should pass unless your changes fail it.

dvstark commented:
I rebased with master and modified the code to explicitly check for optional dependencies. All CI checks have passed

pllim commented:
I just want to let you know that I will be doing a re-review soon. If you have a hard deadline to merge and/or release, please let me know. Thanks!

dvstark commented:
No hard deadline, but certainly will be nice to wrap this up. Thanks in advance!

pllim commented on acstools/findsat_mrt.py:
```diff
@@ -0,0 +1,1362 @@
+        wide kernels included with this package.
+    mask_include_status: list, optional
+        List indicating trails with which status should be considered
+        when making the mask. The default is [2].
```
What does "2" mean? Is this explained somewhere?

dvstark commented:
It’s referring to a status flag (ranging from 0-2) for each trail identified. They are explained in the docstring of filter_sources in utils_findsat_mrt.py, which is definitely too hard to find. I can add an explanation in the TrailFinder class.

--
David V. Stark
Staff Scientist II, Science Support
HST/ACS Team
Space Telescope Science Institute

From: "P. L. Lim" ***@***.***>
Reply-To: spacetelescope/acstools ***@***.***>
Date: Friday, March 24, 2023 at 4:16 PM
To: spacetelescope/acstools ***@***.***>
Cc: David Stark ***@***.***>, Author ***@***.***>
Subject: Re: [spacetelescope/acstools] incorporate findsat_mrt into acstools (PR #176)


@pllim commented on this pull request.

________________________________

In acstools/findsat_mrt.py<https://urldefense.com/v3/__https:/github.com/spacetelescope/acstools/pull/176*discussion_r1148015564__;Iw!!CrWY41Z8OgsX0i-WU-0LuAcUu2o!zSoCVHnOonM6pE1naLlteg-l0sgcytk-RPdKLFiGVISH8JSrKUhubq-bX4ZI2UtjOneMfjb3PbjEulR6ZWaH9W0$>:

> +        Maximum streak width to allow. The default is 75 pixels.

+    buffer : int, optional

+        Size of cutout region extending perpendicular outward from a

+        streak. The default is 250 pixels on each side.

+    threshold : float, optional

+        Minimum S/N when extracting sources from the MRT. The default is 5.

+    theta : ndarray, optional

+        Angles at which to calculate the MRT. The default is None, which

+        sets to np.arange(0,180,0.5).

+    kernels : list, optional

+        Paths to each kernel to be used for source finding in the MRT.

+        The default is None, which reverts to using the 3, 7, and 15 pixel

+        wide kernels included with this package.

+    mask_include_status: list, optional

+        List indicating trails with which status should be considered

+        when making the mask. The default is [2].

What does "2" mean? Is this explained somewhere?

—
Reply to this email directly, view it on GitHub<https://urldefense.com/v3/__https:/github.com/spacetelescope/acstools/pull/176*pullrequestreview-1357366805__;Iw!!CrWY41Z8OgsX0i-WU-0LuAcUu2o!zSoCVHnOonM6pE1naLlteg-l0sgcytk-RPdKLFiGVISH8JSrKUhubq-bX4ZI2UtjOneMfjb3PbjEulR6X3m6ffU$>, or unsubscribe<https://urldefense.com/v3/__https:/github.com/notifications/unsubscribe-auth/ACQIOOEMC6RDRAPT6CKVV4LW5X6JTANCNFSM6AAAAAATTOGSOA__;!!CrWY41Z8OgsX0i-WU-0LuAcUu2o!zSoCVHnOonM6pE1naLlteg-l0sgcytk-RPdKLFiGVISH8JSrKUhubq-bX4ZI2UtjOneMfjb3PbjEulR60cozxn8$>.
You are receiving this because you authored the thread.Message ID: ***@***.***>

pllim reviewed (changes requested):
I have a bunch of proposed changes over at https://github.com/dvstark/acstools/pull/1 . These include API changes and bug fixes as we discussed on Slack. I also applied changes from my earlier rounds of reviews. Reading the diff might drive you crazy, so it might be better for you to view the before/after side by side.

Since these include real bug fixes, if you run the unit test that you wrote with my changes, the result now disagrees with the "truth" you provided here. I cannot tell if that is okay or not, so please look at it carefully.

We can discuss via meeting the week after next if you are unable to review it on your own. Thank you for your continued patience!

pllim commented:
A minor PEP 8 thingy to fix.

And also the link appears broken: https://www.clear.rice.edu/elec431/projects96/DSP/bpanalysis.html

dvstark commented:
I'm suspecting that broken link was temporary; it seems to be working now. I'll fix the pep8 stuff and push again

pllim commented on acstools/findsat_mrt.py:
```diff
@@ -0,0 +1,1409 @@
+    plot : bool, optional
+        Plot all intermediate steps. When set, plots of the input image, MRT
+        with identified sources, and resulting image maks will be generated
+        after running :func:`acstools.findat_mrt.TrailFinder.__init__`:,
```
`__init__` is not exposed in API doc.
```suggestion
        after running the constructor,
```

pllim commented on acstools/findsat_mrt.py:
```diff
@@ -0,0 +1,1409 @@
+        Plot all intermediate steps. When set, plots of the input image, MRT
+        with identified sources, and resulting image maks will be generated
+        after running :func:`acstools.findat_mrt.TrailFinder.__init__`:,
+        :func:`acstools.findsat_mrt.TrailFinder.run_mrt`:,
```
Too many colons.
```suggestion
        :func:`acstools.findsat_mrt.TrailFinder.run_mrt`,
```

pllim commented on acstools/findsat_mrt.py:
```diff
@@ -0,0 +1,1409 @@
+        with identified sources, and resulting image maks will be generated
+        after running :func:`acstools.findat_mrt.TrailFinder.__init__`:,
+        :func:`acstools.findsat_mrt.TrailFinder.run_mrt`:,
+        :func:`acstools.findsat_mrt.TrailFinder.find_mrt_sources`:,
+        :func:`acstools.findsat_mrt.TrailFinder.filter_sources`:,
+        and :func:`acstools.findsat_mrt.TrailFinder.make_mask`:. Users may also
+        generate these plots manually by calling
+        :func:`acstools.findsat_mrt.TrailFinder.plot_image`:,
+        :func:`acstools.findsat_mrt.TrailFinder.plot_mrt`:,
+        :func:`acstools.findsat_mrt.TrailFinder.plot_mask`:, and
+        :func:`acstools.findsat_mrt.TrailFinder.plot_segment`:. The default is
```
```suggestion
        :func:`acstools.findsat_mrt.TrailFinder.find_mrt_sources`,
        :func:`acstools.findsat_mrt.TrailFinder.filter_sources`,
        and :func:`acstools.findsat_mrt.TrailFinder.make_mask`. Users may also
        generate these plots manually by calling
        :func:`acstools.findsat_mrt.TrailFinder.plot_image`,
        :func:`acstools.findsat_mrt.TrailFinder.plot_mrt`,
        :func:`acstools.findsat_mrt.TrailFinder.plot_mask`, and
        :func:`acstools.findsat_mrt.TrailFinder.plot_segment`. The default is
```

pllim commented on acstools/findsat_mrt.py:
```diff
@@ -0,0 +1,1409 @@
+        The default is ``[2]``.
+    plot : bool, optional
+        Plot all intermediate steps. When set, plots of the input image, MRT
+        with identified sources, and resulting image maks will be generated
```
```suggestion
        with identified sources, and resulting image mask will be generated
```

pllim commented on acstools/tests/test_wfc_findsat_mrt.py:
```diff
@@ -0,0 +1,34 @@
+    # compare catalog.
+    WfcWrapper(inputfile, binsize=4, extension=4,
+               output_root=f'{rootname}_flc',
+               output_dir=tmp_path, max_width=75/2.,
```
```suggestion
               output_dir=tmp_path, max_width=37.5,
```

pllim reviewed (commented):
Just some minor comments. I'll approve when they are addressed.

By the way, CI was green after I reran the failed jobs, so the linkcheck and socket failures were indeed transient.

dvstark commented:
All your suggested changes made sense so I committed them all (and sorry if this led to various email notifications from github after every commit)

pllim reviewed (approved):
Thanks! @jryon , do you want to re-review since you requested changes?

If you happen to merge first, please use the `Squash and Merge` button.

jryon reviewed (approved):
I'm happy with it now that @pllim and @dvstark are!

pllim commented:
Merged. Congrats and thanks for your patience! 

@jryon , you are doing releases now, right?

jryon commented:
Yep, added to my to-do list.

dvstark commented:
Thanks both!
